# This script contains the following points:
1. Importing dataset
2. Grouping and Aggregating Data
   * 1 Performing a Single Aggregation on 'df'
   * 2 Performing a Single Aggregation on ords_prods_merge
   * 3 Performing Multiple Aggregations 'df'
3. Creating a new column 'max_order'
4. Assigning loyalty flag
5. Checking the basic statistics of the product prices for each loyalty category
6. Looking at the prices of the items people are buying
7. Determining frequent versus non-frequent customers
8. Export Dataframe

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 1. Importing dataset 

In [4]:
# Project folder path 
path = r'C:\Users\hp\08-2024 Instacart Basket Analysis\Data'
# Import orders products merged
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_derived.pkl'))

In [6]:
# Creating sub set first million rows
df = ords_prods_merge[:1000000]

In [8]:
df.shape

(1000000, 18)

In [12]:
df.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   
2    389851      709             2          0                 21   
3    652770      764             1          3                 13   
4   1813452      764             3          4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest day    busiest days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regularly busy           Most orders  
1  Mid-range product  Regularly busy  Regularly busy        Average orders  
2  Mid-range product     Busiest day    Busiest days         Fewest orders  
3  Mid-range product  Regularly busy    Slowest days           Most orders  
4  Mid-range product      Least busy    Slowest days        Average orders

# 2. Grouping and Aggregating Data 

In [121]:
#Trying the function groupby on the sub set 
df.groupby('product_name')
#--> First step completed by spliting the data into groups  (with the group being the “product_name” column).

### 2.1 Performing a Single Aggregation on 'df'

In [193]:
#Apply the agg() function to each group to obtain the mean values for the “order_number” column.
df.groupby('department_id').agg({'order_number': ['mean']})

#ways : 
#df.groupby('department_id')['order_number'].mean()
#df.groupby('department_id').order_number.mean()

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

### 2.2 Performing a Single Aggregation on ords_prods_merge 

In [110]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.666130
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#### Insights:
The subset data generally approximates the entire dataset, but there are a few departments (notably 21, 10, and 12) where the differences are more pronounced. For more accurate insights the full dataset is preferred.

### 2.3 Performing Multiple Aggregations 'df'

In [35]:
#Performing Multiple Aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.577493   1  99
2                17.320781   1  96
3                16.084944   1  99
4                17.530458   1  99
5                14.763075   1  99
6                16.658449   1  99
7                17.031590   1  99
8                15.076662   1  98
9                15.447580   1  99
10               18.681852   1  99
11               15.447411   1  99
12               14.327957   1  99
13               16.548642   1  99
14               16.960241   1  99
15               16.121948   1  99
16               17.803851   1  99
17               15.593633   1  99
18               19.674252   1  99
19               16.899756   1  99
20               16.255442   1  99
21               25.535479   1  98

# 3. Creating a new column 'max_order'

In [108]:
# Creating a new column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform('max')
#Check the dataset 
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    3139998      138            28          6                 11   
1    1977647      138            30          6                 17   
2     389851      709             2          0                 21   
3     652770      764             1          3                 13   
4    1813452      764             3          4                 17   
5    1701441      777            16          1                  7   
6    1871483      825             3          2                 14   
7    1290456      910            12          3                 10   
8     369558     1052            10          1                 20   
9     589712     1052            15          1                 12   
10   1798115     1379             1          5                 11   
11    497193     1480            36          3                 13   
12   1727826     1480            38          4                 13   
13   2694698     1480            39          2                 12   
14   3171753     1494             6          3                 10   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     NaN                  1          0   both   
11                     7.0                  1          0   both   
12                     7.0                  3          1   both   
13                    12.0                  4          1   both   
14                     2.0                  8          0   both   

      price_range_loc     busiest day    busiest days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy           Most orders   
1   Mid-range product  Regularly busy  Regularly busy        Average orders   
2   Mid-range product     Busiest day    Busiest days         Fewest orders   
3   Mid-range product  Regularly busy    Slowest days           Most orders   
4   Mid-range product      Least busy    Slowest days        Average orders   
5   Mid-range product  Regularly busy    Busiest days         Fewest orders   
6   Mid-

In [127]:
#Check the shape
ords_prods_merge.shape

(32404161, 20)

In [129]:
#To look at the whole result
pd.options.display.max_rows = None

# 4. Assigning loyalty flag 

In [96]:
#create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value.
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [98]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [100]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [102]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876363
Loyal customer      10284010
New customer         6243788
Name: count, dtype: int64

#### Insights: 
Regular Customers make up the largest group, representing the bulk of the customer base.

Loyal Customers form a significant portion, indicating a strong base of dedicated customers.

New Customers are a smaller, but important group, representing potential growth and future loyal customers.

In [133]:
# Check Loyalty flag in dataframe
ords_prods_merge[['user_id', 'loyalty_flag','order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845  Regular customer             6
59     6866      New customer             1

# 5. checking the basic statistics of the product prices for each loyalty category

In [137]:
# We’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer     9.579349  1.0  14900.0
New customer      10.059917  1.0  14900.0
Regular customer   9.894716  1.0  14900.0

#### Analysis:
1. Mean Prices:
  * Loyal Customers have the lowest average purchase price at 9.58.
  * New Customers have the highest average purchase price at 10.06.
  * Regular Customers fall in between, with an average price of 9.89.

2. Min and Max Prices:
 * The minimum and maximum prices are identical across all customer segments, indicating that all customer groups have access to the same price range for products.  
##### There is a small but noticeable difference in the average prices of products purchased by different customer groups. New customers tend to spend more, while loyal customers spend slightly less. If your goal is to increase the average purchase price, you might want to explore targeted strategies for regular and loyal customers, such as introducing premium product offerings or upselling strategies.

# 6. Looking at the prices of the items people are buying

In [161]:
# Creating column for average spender
ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [163]:
# Creating flag for low and high spender
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [165]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     31827370
High_spender      576791
Name: count, dtype: int64

#### Insights
The vast majority of customers fall into the Low Spender category. A much smaller group of customers are classified as High Spenders, likely contributing significantly more revenue per customer compared to Low Spenders. This distribution is common in many businesses, where a small percentage of customers (High Spenders) often generate a large portion of the revenue, while the majority (Low Spenders) contribute smaller amounts individually. 

**Suggestions:** 
1. High Spender Focus: Although the High Spender group is much smaller, they likely contribute disproportionately to the overall revenue. Targeting this group with personalized offers, loyalty programs, or premium products could be beneficial.
2. Low Spender Engagement: For Low Spenders, strategies could include upselling, cross-selling, or increasing the frequency of purchases to boost their spending.

# 7. determining frequent versus non-frequent customers

In [174]:
# Create a column for frequet users
ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [176]:
# Creating flag on order frequency 
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'

In [179]:
ords_prods_merge.loc[(ords_prods_merge['Customer_frequency']>10) & (ords_prods_merge['Customer_frequency']<=20),'frequency_flag'] = 'Regular customer'

In [181]:
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [183]:
# Check frequency flags
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        21559380
Regular customer          7208433
Non_frequent_customer     3636343
NaN                             5
Name: count, dtype: int64

--> NaN means cutomers who bought once 

In [191]:
ords_prods_merge[['user_id','days_since_prior_order', 'Customer_frequency', 'frequency_flag']].head(20)

user_id  days_since_prior_order  Customer_frequency     frequency_flag
0       138                     3.0                 8.0  Frequent Customer
1       138                    20.0                 8.0  Frequent Customer
2       709                     6.0                 8.0  Frequent Customer
3       764                     NaN                 9.0  Frequent Customer
4       764                     9.0                 9.0  Frequent Customer
5       777                    26.0                11.0   Regular customer
6       825                    30.0                20.0   Regular customer
7       910                    30.0                 6.0  Frequent Customer
8      1052                    19.0                10.0  Frequent Customer
9      1052                    15.0                10.0  Frequent Customer
10     1379                     NaN                20.0   Regular customer
11     1480                     7.0                 7.0  Frequent Customer
12     1480                     7.0                 7.0  Frequent Customer
13     1480                    12.0                 7.0  Frequent Customer
14     1494                     2.0                 7.0  Frequent Customer
15     1494                     6.0                 7.0  Frequent Customer
16     1494                    10.0                 7.0  Frequent Customer
17     1540                     7.0                 7.0  Frequent Customer
18     1540                     6.0                 7.0  Frequent Customer
19     1540                    14.0                 7.0  Frequent Customer

In [187]:
#Checking the dataset : 
ords_prods_merge.shape 

(32404161, 24)

In [189]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  ...  \
0   3139998      138            28          6                 11  ...   
1   1977647      138            30          6                 17  ...   
2    389851      709             2          0                 21  ...   
3    652770      764             1          3                 13  ...   
4   1813452      764             3          4                 17  ...   

     price_range_loc     busiest day    busiest days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most orders   
1  Mid-range product  Regularly busy  Regularly busy        Average orders   
2  Mid-range product     Busiest day    Busiest days         Fewest orders   
3  Mid-range product  Regularly busy    Slowest days           Most orders   
4  Mid-range product      Least busy    Slowest days        Average orders   

  max_order      loyalty_flag average_spend spender_flag  Customer_frequency  \
0        32  Regular customer      6.935811  Low_spender                 8.0   
1        32  Regular customer      6.935811  Low_spender                 8.0   
2         5      New customer      7.930208  Low_spender                 8.0   
3         3      New customer      4.972414  Low_spender                 9.0   
4         3      New customer      4.972414  Low_spender                 9.0   

      frequency_flag  
0  Frequent Customer  
1  Frequent Customer  
2  Frequent Customer  
3  Frequent Customer  
4  Frequent Customer  

[5 rows x 24 columns]

# 8. Export Dataframe: 

In [199]:
ords_prods_merge.to_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_aggregated.pkl'))